finetuner/utils.py

```diff
@@ -0,0 +1,198 @@
+        in the CSV
+
+    """
+
+    if hasattr(file, 'read'):
+        file_ctx = nullcontext(file)
+    else:
+        file_ctx = open(file, 'r', encoding=options.encoding)
+
+    with file_ctx as fp:
+        # when set to auto, then sniff
+        try:
+            if isinstance(options.dialect, str) and options.dialect == 'auto':
+                options.dialect = csv.Sniffer().sniff(fp.read(1024))
+                fp.seek(0)
+        except Exception:
+            options.dialect = 'excel'  #: can not sniff delimiter, use default dialect
+
+        lines = csv.reader(fp, dialect=options.dialect)
+
+        artificial_label = 0
+        t1, t2 = None, None
+
+        for col1, col2 in _subsample(lines, options.size, options.sampling_rate):
+            if not t1:  # determining which column contains images
+                t1, t2 = check_columns(task, col1, col2)
+
+            d1 = create_document(t1, col1, options.convert_to_blob)
+
+            if options.is_labeled:
+                label = col2
+                d2 = None
+            else:
+                d2 = create_document(t2, col2, options.convert_to_blob)
+
+            if d2 is None:
+                d1.tags[DEFAULT_TAG_KEY] = label
+                yield d1
+            elif (d1.text and d2.text) or (d1.uri and d2.uri):
+                # same modality
+                d1.tags[DEFAULT_TAG_KEY] = artificial_label
+                d2.tags[DEFAULT_TAG_KEY] = artificial_label
+                artificial_label += 1
+                yield d1
+                yield d2
+            else:
+                # different modalities, for CLIP
+                d1.modality = t1
+                d2.modality = t2
+                yield Document(chunks=[d1, d2])
+
+
+def check_columns(
+    task: str,
+    col1: str,
+    col2: str,
+) -> Tuple[str, str]:
+    """Determines the expected modalities of each column using the task argument,
+        Then checks the given row of the CSV to confirm that it contains valid data
+
+    :param task: The task of the model being used.
+    :param col1: A single value from the first column of the CSV.
+    :param col2: A single value from the second column of the CSV.
+    :return: The expected modality of each column
+    """
+
+    if task == 'any':
+        raise ValueError('MLP model does not support values read in from CSV files.')
+
+    if len(task.split('-to-')) == 2:
+        t1, t2 = task.split('-to-')
+    else:
+        raise ValueError(f'Model has invalid task: {task}')
+
+    if t1 == 'text' and t2 == 'image':
+        if _is_uri(col1) and not _is_uri(col2):
+            t1 = 'image'
+            t2 = 'text'
+        elif not _is_uri(col2):
+            raise ValueError(
+                (
+                    'uri required in at least one colum ',
+                    'for model with task: ',
+                    t1,
+                    '-to-',
+                    t2,
+                    '.',
```

Comment:
Uri required, also is it working? i guess you need f-string it?

Reply:
ah, it did work but it was as a tuple, not a single string. Fixed now